ndb/src/google/cloud/ndb/tasklets.py

```diff
@@ -44,18 +49,77 @@ def add_flow_exception(*args, **kwargs):
 
 
 class Future:
-    __slots__ = ()
+    """Base class for all NDB futures.
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    Provides interface defined by :class:`concurrent.futures.Future` as well as
+    legacy NDB ``Future`` class.
+
+    Sketchy code, very incomplete.
+    """
+    def __init__(self):
+        self._done = False
+        self._result = None
+
+    def set_result(self, result):
+        self._result = result
+        self._done = True
+
+    def done(self):
+        return self._done
+
+    def result(self):
+        while not self._done:
+            _eventloop.run1()
+        return self._result
+
+
+class TaskletFuture(Future):
+    """A Future for an NDB tasklet.
+
+    Sketchy code, very incomplete.
+    """
+    def __init__(self, generator):
+        super(TaskletFuture, self).__init__()
+        self.generator = generator
+
+    def _advance_tasklet(self, send_value=None):
+        """Advances tasklet one step given a value to send into generator."""
+        try:
+            # Send next value in to generator (will be None if just starting)
+            yielded = self.generator.send(send_value)
+
+        except StopIteration as stop:
+            # Tasklet has finished
+            self.set_result(get_return_value(stop))
```

Comment:
Surprised to see the exception as being useful

Reply:
Well, a tasklet is an iterator, so you know it's done when it raises StopIteration.

If a generator function returns a value, that's returned with the StopIteration exception. Legacy also defined a ``Return`` exception so people could do ``raise Return(foo)`` to return a value from a tasklet. My guess is earlier versions of Python didn't support returning from a generator. ``Return`` is no longer necessary but likely to pop up in legacy code, so it gets to stay! 

From our point of view, there's no real distinction between a tasklet just returning a value like a function and raising Return, as it all winds up with us digging the return value out of the StopIteration exception.